File: reference/task-facts/1.0/evidence/interop/cycles/cycle-1/implementation-python/evaluate.py
"""Independent implementation of the frozen Task Facts 0.2 contract."""
import datetime as dt
import hashlib
import json
import re
import sys
from pathlib import Path

class Invalid(Exception):
    def __init__(self, reason):
        self.reason = reason
        super().__init__(reason)

SCHEMA = json.loads(Path(__file__).with_name('task-facts.schema.json').read_text())

def tfj(value):
    if value is None or type(value) is bool:
        return
    if type(value) is int:
        if abs(value) <= 9007199254740991:
            return
    elif type(value) is str:
        if not any(0xD800 <= ord(c) <= 0xDFFF for c in value):
            return
    elif type(value) is list:
        for item in value:
            tfj(item)
        return
    elif type(value) is dict:
        for key, item in value.items():
            if type(key) is not str:
                raise Invalid('NON_CANONICAL_INPUT')
            tfj(key)
            tfj(item)
        return
    raise Invalid('NON_CANONICAL_INPUT')

def canonical_bytes(value):
    tfj(value)
    return json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(',', ':')).encode('utf-8')

def digest(value):
    return 'sha256:' + hashlib.sha256(canonical_bytes(value)).hexdigest()

snapshot_hash = digest

def strict_load(raw):
    def pairs(items):
        out = {}
        for k, v in items:
            if k in out:
                raise ValueError('duplicate decoded key')
            out[k] = v
        return out
    def constant(value):
        raise ValueError('non-JSON constant')
    try:
        text = raw.decode('utf-8') if isinstance(raw, bytes) else raw
        # Oversized tokens remain JSON integers; classify them only after parsing.
        def integer(token):
            return int(token) if len(token.lstrip('-')) <= 16 else 9007199254740992
        result = json.loads(text, object_pairs_hook=pairs, parse_constant=constant, parse_int=integer)
    except (ValueError, UnicodeError):
        raise Invalid('JSON_PARSE_ERROR') from None
    tfj(result)
    return result

def schema_valid(value, schema):
    if '$ref' in schema:
        return schema_valid(value, SCHEMA['$defs'][schema['$ref'].split('/')[-1]])
    if 'oneOf' in schema:
        return sum(schema_valid(value, s) for s in schema['oneOf']) == 1
    if 'const' in schema and canonical_bytes(value) != canonical_bytes(schema['const']):
        return False
    if 'enum' in schema and not any(canonical_bytes(value) == canonical_bytes(v) for v in schema['enum']):
        return False
    typ = schema.get('type')
    types = {'object': dict, 'array': list, 'string': str, 'integer': int, 'boolean': bool}
    if typ and type(value) is not types[typ]:
        return False
    if type(value) is str:
        if len(value) < schema.get('minLength', 0):
            return False
        if 'pattern' in schema and not re.search(schema['pattern'], value):
            return False
    if type(value) is list:
        if len(value) < schema.get('minItems', 0):
            return False
        if schema.get('uniqueItems') and len({canonical_bytes(v) for v in value}) != len(value):
            return False
        if not all(schema_valid(v, schema.get('items', {})) for v in value):
            return False
    if type(value) is dict:
        if not set(schema.get('required', [])).issubset(value):
            return False
        props = schema.get('properties', {})
        for k, v in value.items():
            if not schema_valid(k, schema.get('propertyNames', {})):
                return False
            rule = props.get(k, schema.get('additionalProperties', True))
            if rule is False or (isinstance(rule, dict) and not schema_valid(v, rule)):
                return False
    return True

def validate(value, name):
    if not schema_valid(value, SCHEMA['$defs'][name]):
        raise Invalid('SCHEMA_INVALID')

def temporal(value, typ):
    if type(value) is not str:
        return None
    try:
        if typ == 'date' and re.fullmatch(r'[0-9]{4}-[0-9]{2}-[0-9]{2}', value):
            return dt.date.fromisoformat(value)
        if typ == 'date-time' and re.fullmatch(r'[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(Z|[+-][0-9]{2}:[0-9]{2})', value):
            zone = value[19:]
            if zone != 'Z':
                hh, mm = int(zone[1:3]), int(zone[4:])
                if zone == '-00:00' or hh > 14 or mm > 59 or (hh == 14 and mm != 0):
                    return None
            return dt.datetime.fromisoformat(value.replace('Z', '+00:00'))
    except ValueError:
        pass
    return None

def valid_value(value, typ):
    if typ == 'string':
        return type(value) is str and len(value) > 0
    if typ == 'boolean':
        return type(value) is bool
    if typ == 'integer':
        return type(value) is int and abs(value) <= 9007199254740991
    if typ in ('date', 'date-time'):
        return temporal(value, typ) is not None
    if typ in ('string-set', 'association-set'):
        if type(value) is not list or not value:
            return False
        if typ == 'string-set':
            good = all(valid_value(v, 'string') for v in value)
        else:
            good = all(type(v) is list and len(v) == 2 and all(valid_value(x, 'string') for x in v) for v in value)
        return good and len({canonical_bytes(v) for v in value}) == len(value)
    return False

def validate_payload(snapshot, condition, context):
    validate(snapshot, 'snapshot')
    declarations, facts, evidence = (snapshot[k] for k in ('declarations', 'facts', 'evidence'))
    if set(facts) - set(declarations):
        raise Invalid('UNDECLARED_FACT')
    for key in sorted(facts):
        fact = facts[key]
        if fact['state'] == 'known' and not valid_value(fact['value'], declarations[key]['type']):
            raise Invalid('TYPE_MISMATCH')
    used = set()
    for key, fact in facts.items():
        for ref in fact['evidenceRefs']:
            if ref not in evidence or evidence[ref]['fact'] != key:
                raise Invalid('EVIDENCE_REFERENCE')
            used.add(ref)
    if used != set(evidence):
        raise Invalid('EVIDENCE_REFERENCE')
    validate(context, 'verificationContext')
    validate(condition, 'condition')
    for atom in condition['all']:
        if atom['fact'] not in declarations:
            raise Invalid('UNDECLARED_FACT')
        typ = declarations[atom['fact']]['type']
        op = atom['op']
        if op in ('present', 'absent'):
            continue
        value = atom['value']
        if op == 'contains':
            if typ not in ('string-set', 'association-set'):
                raise Invalid('OPERATOR_TYPE')
            if not valid_value([value], typ):
                raise Invalid('TYPE_MISMATCH')
        else:
            if typ in ('string-set', 'association-set'):
                raise Invalid('OPERATOR_TYPE')
            if typ in ('date', 'date-time'):
                good = valid_value(value, 'date') or valid_value(value, 'date-time')
            else:
                good = valid_value(value, typ)
            if not good:
                raise Invalid('TYPE_MISMATCH')

def atom_result(snapshot, atom, context, hash_value):
    key = atom['fact']
    fact = snapshot['facts'].get(key)
    if fact is None:
        return 'UNKNOWN', 'FACT_MISSING'
    task = snapshot['task']
    identity = {'taskId': task['id'], 'artifactHash': task['artifactHash'], 'inputPackageHash': task['inputPackageHash']}
    verified = context['snapshotHash'] == hash_value and all(context[k] == v for k, v in identity.items())
    assertion_hash = digest({k: v for k, v in fact.items() if k != 'evidenceRefs'})
    for ref in fact['evidenceRefs']:
        ev = snapshot['evidence'][ref]
        verified = verified and ev['status'] == 'verified' and context['acceptedEvidence'].get(ref) == digest(ev) and ev['assertionHash'] == assertion_hash and all(ev[k] == v for k, v in identity.items())
    if not verified:
        return 'UNKNOWN', 'EVIDENCE_UNVERIFIED'
    state = fact['state']
    if state == 'unknown':
        return 'UNKNOWN', 'FACT_UNKNOWN'
    if state == 'contradictory':
        return 'CONTRADICTORY', 'FACT_CONTRADICTORY'
    typ = snapshot['declarations'][key]['type']
    if state == 'known' and typ == 'association-set':
        sources = {}
        for source, target in fact['value']:
            if source in sources and sources[source] != target:
                return 'CONTRADICTORY', 'ASSOCIATION_CONFLICT'
            sources[source] = target
    op = atom['op']
    if op == 'present':
        truth = state == 'known'
    elif op == 'absent':
        truth = state == 'empty'
    elif state == 'empty':
        truth = False
    elif op == 'contains':
        truth = atom['value'] in fact['value']
    elif typ in ('date', 'date-time'):
        other_typ = 'date' if temporal(atom['value'], 'date') is not None else 'date-time'
        if other_typ != typ:
            return 'UNKNOWN', 'TIME_BASIS_UNRESOLVED'
        truth = temporal(fact['value'], typ) == temporal(atom['value'], typ)
    else:
        truth = fact['value'] == atom['value']
    return ('APPLIES', 'ALL_TRUE') if truth else ('DOES_NOT_APPLY', 'PREDICATE_FALSE')

def evaluate(snapshot, condition, verification_context):
    record = {'conditionId': None, 'snapshotHash': None, 'outcome': 'INVALID', 'reason': None}
    try:
        tfj([snapshot, condition, verification_context])
        validate(condition, 'routingCondition')
        record.update(conditionId=condition['id'], snapshotHash=digest(snapshot))
        validate_payload(snapshot, condition, verification_context)
        results = [atom_result(snapshot, atom, verification_context, record['snapshotHash']) for atom in condition['all']]
        ranks = {'APPLIES': 0, 'DOES_NOT_APPLY': 1, 'UNKNOWN': 2, 'CONTRADICTORY': 3}
        outcome = max((r[0] for r in results), key=ranks.get)
        reason = min(r[1] for r in results if r[0] == outcome)
        record.update(outcome=outcome, reason=reason)
    except Invalid as error:
        record['reason'] = error.reason
    return record

def unbound(reason):
    return dict(family=None, caseId=None, conditionId=None, snapshotHash=None, outcome='INVALID', reason=reason)

def evaluate_family(family):
    tfj(family)
    validate(family, 'routingFamily')
    cases = family['cases']
    if len({c['id'] for c in cases}) != len(cases):
        raise Invalid('DUPLICATE_ID')
    for case in cases:
        if len({c['id'] for c in case['conditions']}) != len(case['conditions']):
            raise Invalid('DUPLICATE_ID')
    results = []
    for case in cases:
        hash_value = digest(case['snapshot'])
        for condition in case['conditions']:
            record = dict(family=family['family'], caseId=case['id'], conditionId=condition['id'], snapshotHash=hash_value, outcome='INVALID', reason=None)
            try:
                validate(family, 'family')
                validate(case, 'case')
                record.update(evaluate(case['snapshot'], condition, case['verificationContext']))
            except Invalid as error:
                record['reason'] = error.reason
            results.append(record)
    return results

def evaluate_path(path):
    try:
        try:
            raw = Path(path).read_bytes()
        except OSError:
            raise Invalid('INPUT_IO_ERROR') from None
        return evaluate_family(strict_load(raw))
    except Invalid as error:
        return [unbound(error.reason)]

def main(argv=None):
    paths = sys.argv[1:] if argv is None else argv
    results = [r for path in paths for r in evaluate_path(path)]
    print(json.dumps({'results': results}, ensure_ascii=False, indent=2))
    return 2 if any(r['outcome'] == 'INVALID' for r in results) else 0

if __name__ == '__main__':
    sys.exit(main())

File: reference/task-facts/1.0/evidence/interop/cycles/cycle-1/implementation-node/evaluate.mjs
import fs from 'node:fs';
import crypto from 'node:crypto';
import {fileURLToPath} from 'node:url';
export class Invalid extends Error { constructor(reason) {super(reason);this.reason=reason;} }
const fail = reason => {throw new Invalid(reason);};
const own=(o,k)=>Object.hasOwn(o,k);
const object=x=>x!==null && typeof x==='object' && !Array.isArray(x);
export function parseStrict(bytes) {
  let s; try {s=new TextDecoder('utf-8',{fatal:true,ignoreBOM:true}).decode(bytes);} catch {fail('JSON_PARSE_ERROR');}
  let i=0, noncanonical=false;
  const ws=()=>{while(/[\x20\t\r\n]/.test(s[i]??'!'))i++;};
  function string() {
    const start=i++; let closed=false;
    while(i<s.length) {const c=s[i++]; if(c==='"'){closed=true;break;} if(c==='\\'){if(i>=s.length)break;const e=s[i++];if(e==='u'){if(!/^[0-9a-fA-F]{4}$/.test(s.slice(i,i+4)))fail('JSON_PARSE_ERROR');i+=4;}else if(!'"\\/bfnrt'.includes(e))fail('JSON_PARSE_ERROR');}else if(c.charCodeAt(0)<32)fail('JSON_PARSE_ERROR');}
    if(!closed)fail('JSON_PARSE_ERROR');try{return JSON.parse(s.slice(start,i));}catch{fail('JSON_PARSE_ERROR');}
  }
  function value(){ws();const c=s[i];if(c==='"')return string();
    if(c==='{'){i++;ws();const o=Object.create(null);if(s[i]==='}'){i++;return o;}while(true){ws();if(s[i]!=='"')fail('JSON_PARSE_ERROR');const k=string();if(own(o,k))fail('JSON_PARSE_ERROR');ws();if(s[i++]!==':')fail('JSON_PARSE_ERROR');o[k]=value();ws();const d=s[i++];if(d==='}')return o;if(d!==',')fail('JSON_PARSE_ERROR');}}
    if(c==='['){i++;ws();const a=[];if(s[i]===']'){i++;return a;}while(true){a.push(value());ws();const d=s[i++];if(d===']')return a;if(d!==',')fail('JSON_PARSE_ERROR');}}
    for(const [token,v] of [['true',true],['false',false],['null',null]])if(s.startsWith(token,i)){i+=token.length;return v;}
    const m=s.slice(i).match(/^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?/);if(!m)fail('JSON_PARSE_ERROR');i+=m[0].length;
    if(/[.eE]/.test(m[0]))noncanonical=true;else if(BigInt(m[0])>9007199254740991n || BigInt(m[0])< -9007199254740991n)noncanonical=true;
    return Number(m[0]);
  }
  const result=value();ws();if(i!==s.length)fail('JSON_PARSE_ERROR');if(noncanonical)fail('NON_CANONICAL_INPUT');validateTFJ(result);return result;
}
export function validateTFJ(x,seen=new Set()) {
  if(x===null || typeof x==='boolean')return;
  if(typeof x==='number'){if(!Number.isSafeInteger(x))fail('NON_CANONICAL_INPUT');return;}
  if(typeof x==='string'){for(const c of x){const n=c.codePointAt(0);if(n>=0xd800 && n<=0xdfff)fail('NON_CANONICAL_INPUT');}return;}
  if(!Array.isArray(x)&&!object(x))fail('NON_CANONICAL_INPUT');if(seen.has(x))fail('NON_CANONICAL_INPUT');seen.add(x);
  if(Array.isArray(x)){for(let i=0;i<x.length;i++)validateTFJ(x[i],seen);}else{if(![Object.prototype,null].includes(Object.getPrototypeOf(x)))fail('NON_CANONICAL_INPUT');for(const k of Object.keys(x)){validateTFJ(k,seen);validateTFJ(x[k],seen);}}seen.delete(x);
}
export function scalarCompare(a,b){const aa=Array.from(a,c=>c.codePointAt(0)),bb=Array.from(b,c=>c.codePointAt(0));for(let i=0;i<Math.min(aa.length,bb.length);i++)if(aa[i]!==bb[i])return aa[i]-bb[i];return aa.length-bb.length;}
export function canonical(x){validateTFJ(x);function emit(v){if(v===null||typeof v!=='object')return JSON.stringify(v);if(Array.isArray(v))return '['+v.map(emit).join(',')+']';return '{'+Object.keys(v).sort(scalarCompare).map(k=>JSON.stringify(k)+':'+emit(v[k])).join(',')+'}';}return emit(x);}
export const hash=x=>'sha256:'+crypto.createHash('sha256').update(canonical(x),'utf8').digest('hex');
const schema=JSON.parse(fs.readFileSync(new URL('./schemas/task-facts.schema.json',import.meta.url),'utf8'));
// Implements the schema vocabulary used by the supplied frozen schema.
function matches(x,s){if(s.$ref)return matches(x,schema.$defs[s.$ref.split('/').at(-1)]);if(s.oneOf&&s.oneOf.filter(v=>matches(x,v)).length!==1)return false;
if(own(s,'const')&&canonical(x)!==canonical(s.const))return false;if(s.enum&&!s.enum.some(v=>canonical(x)===canonical(v)))return false;
if(s.type && !({object:object(x),array:Array.isArray(x),string:typeof x==='string',boolean:typeof x==='boolean',integer:Number.isSafeInteger(x)})[s.type])return false;
if(typeof x==='string'&&((s.minLength!==undefined&&Array.from(x).length<s.minLength)||(s.pattern&&!new RegExp(s.pattern).test(x))))return false;
if(Array.isArray(x)){if(s.minItems!==undefined&&x.length<s.minItems)return false;if(s.uniqueItems&&new Set(x.map(canonical)).size!==x.length)return false;if(s.items&&!x.every(v=>matches(v,s.items)))return false;}
if(object(x)){if(s.required&&!s.required.every(k=>own(x,k)))return false;for(const k of Object.keys(x)){if(s.propertyNames&&!matches(k,s.propertyNames))return false;if(s.properties&&own(s.properties,k)){if(!matches(x[k],s.properties[k]))return false;}else if(s.additionalProperties===false)return false;else if(object(s.additionalProperties)&&!matches(x[k],s.additionalProperties))return false;}}return true;}
const check=(x,name)=>{if(!matches(x,schema.$defs[name]))fail('SCHEMA_INVALID');};
function date(v){if(typeof v!=='string'||v.length!==10||!/^\d{4}-\d{2}-\d{2}$/.test(v))return false;const [y,m,d]=v.split('-').map(Number);return y>=1&&m>=1&&m<=12&&d>=1&&d<=([31,(y%4===0&&(y%100!==0||y%400===0))?29:28,31,30,31,30,31,31,30,31,30,31][m-1]);}
function instant(v){if(typeof v!=='string')return false;const m=v.match(/^(\d{4}-\d{2}-\d{2})T(\d{2}):(\d{2}):(\d{2})(Z|([+-])(\d{2}):(\d{2}))$/);return !!m&&m[0].length===v.length&&date(m[1])&&+m[2]<24&&+m[3]<60&&+m[4]<60&&m[5]!=='-00:00'&&(m[5]==='Z'||(+m[7]<=14&&+m[8]<60&&(+m[7]!==14||+m[8]===0)));}
const text=x=>typeof x==='string'&&x.length>0;
const pair=x=>Array.isArray(x)&&x.length===2&&x.every(text);
function typed(x,t){switch(t){case 'string':return text(x);case 'boolean':return typeof x==='boolean';case 'integer':return Number.isSafeInteger(x);case 'date':return date(x);case 'date-time':return instant(x);case 'string-set':case 'association-set':return Array.isArray(x)&&x.length>0&&x.every(t==='string-set'?text:pair)&&new Set(x.map(canonical)).size===x.length;default:return false;}}
function validateBound(s,c,v){check(s,'snapshot');for(const k of Object.keys(s.facts))if(!own(s.declarations,k))fail('UNDECLARED_FACT');for(const k of Object.keys(s.facts).sort(scalarCompare)){const f=s.facts[k];if(f.state==='known'&&!typed(f.value,s.declarations[k].type))fail('TYPE_MISMATCH');}
const used=new Set();for(const [k,f] of Object.entries(s.facts))for(const r of f.evidenceRefs){if(!own(s.evidence,r)||s.evidence[r].fact!==k)fail('EVIDENCE_REFERENCE');used.add(r);}for(const k of Object.keys(s.evidence))if(!used.has(k))fail('EVIDENCE_REFERENCE');check(v,'verificationContext');check(c,'condition');
for(const a of c.all){if(!own(s.declarations,a.fact))fail('UNDECLARED_FACT');const t=s.declarations[a.fact].type;if(a.op==='equals'){if(t.endsWith('-set'))fail('OPERATOR_TYPE');if(t==='date'||t==='date-time'){if(!date(a.value)&&!instant(a.value))fail('TYPE_MISMATCH');}else if(!typed(a.value,t))fail('TYPE_MISMATCH');}else if(a.op==='contains'){if(!t.endsWith('-set'))fail('OPERATOR_TYPE');if(!(t==='string-set'?text(a.value):pair(a.value)))fail('TYPE_MISMATCH');}}
}
function atom(s,a,v,h){const f=s.facts[a.fact];if(!f)return ['UNKNOWN','FACT_MISSING'];const t=s.task;let verified=v.snapshotHash===h&&v.taskId===t.id&&v.artifactHash===t.artifactHash&&v.inputPackageHash===t.inputPackageHash;const assertion={...f};delete assertion.evidenceRefs;
for(const r of f.evidenceRefs){const e=s.evidence[r];verified=verified&&e.status==='verified'&&v.acceptedEvidence[r]===hash(e)&&e.assertionHash===hash(assertion)&&e.taskId===t.id&&e.artifactHash===t.artifactHash&&e.inputPackageHash===t.inputPackageHash;}
if(!verified)return ['UNKNOWN','EVIDENCE_UNVERIFIED'];if(f.state==='unknown')return ['UNKNOWN','FACT_UNKNOWN'];if(f.state==='contradictory')return ['CONTRADICTORY','FACT_CONTRADICTORY'];const type=s.declarations[a.fact].type;
if(f.state==='known'&&type==='association-set'){const m=new Map();for(const [source,target] of f.value){if(m.has(source)&&m.get(source)!==target)return ['CONTRADICTORY','ASSOCIATION_CONFLICT'];m.set(source,target);}}
let yes=false;if(a.op==='present')yes=f.state==='known';else if(a.op==='absent')yes=f.state==='empty';else if(f.state==='known'){if(a.op==='contains')yes=f.value.some(x=>canonical(x)===canonical(a.value));else if(type==='date'||type==='date-time'){if(date(f.value)!==date(a.value))return ['UNKNOWN','TIME_BASIS_UNRESOLVED'];yes=type==='date'?f.value===a.value:Date.parse(f.value)===Date.parse(a.value);}else yes=f.value===a.value;}
return yes?['APPLIES','ALL_TRUE']:['DOES_NOT_APPLY','PREDICATE_FALSE'];}
export function evaluate(s,c,v){let h=null;let id=null;try{validateTFJ([s,c,v]);if(!object(c)||!matches(c.id??null,schema.$defs.identifier))fail('SCHEMA_INVALID');id=c.id;h=hash(s);validateBound(s,c,v);const rank={APPLIES:0,DOES_NOT_APPLY:1,UNKNOWN:2,CONTRADICTORY:3};const r=c.all.map(a=>atom(s,a,v,h)).sort((a,b)=>rank[b[0]]-rank[a[0]]||scalarCompare(a[1],b[1]))[0];return {conditionId:id,snapshotHash:h,outcome:r[0],reason:r[1]};}catch(e){if(!(e instanceof Invalid))throw e;return {conditionId:id,snapshotHash:h,outcome:'INVALID',reason:e.reason};}}
const unbound=reason=>({family:null,caseId:null,conditionId:null,snapshotHash:null,outcome:'INVALID',reason});
export function evaluate_family(f){validateTFJ(f);check(f,'routingFamily');if(new Set(f.cases.map(c=>c.id)).size!==f.cases.length)fail('DUPLICATE_ID');for(const c of f.cases)if(new Set(c.conditions.map(a=>a.id)).size!==c.conditions.length)fail('DUPLICATE_ID');const out=[];for(const c of f.cases)for(const a of c.conditions){let r;try{check(f,'family');check(c,'case');r=evaluate(c.snapshot,a,c.verificationContext);}catch(e){if(!(e instanceof Invalid))throw e;r={conditionId:a.id,snapshotHash:hash(c.snapshot),outcome:'INVALID',reason:e.reason};}out.push({family:f.family,caseId:c.id,...r});}return out;}
export function evaluate_path(path){let raw;try{raw=fs.readFileSync(path);}catch{return [unbound('INPUT_IO_ERROR')];}try{return evaluate_family(parseStrict(raw));}catch(e){if(!(e instanceof Invalid))throw e;return [unbound(e.reason)];}}
export function main(paths){const results=paths.flatMap(evaluate_path);process.stdout.write(JSON.stringify({results},null,2)+'\n');return results.some(r=>r.outcome==='INVALID')?2:0;}
if(process.argv[1]&&fileURLToPath(import.meta.url)===fs.realpathSync(process.argv[1]))process.exitCode=main(process.argv.slice(2));
